bdk_core/src/keychain.rs

```diff
@@ -0,0 +1,56 @@
+use crate::{
+    chain_graph::{self, ChainGraph},
+    collections::BTreeMap,
+};
+
+#[derive(Clone, Debug, PartialEq)]
+/// An update that includes the last active indexes of each keychain.
+pub struct KeychainScan<K, I> {
+    /// The update data in the form of a chain that could be applied
+    pub update: ChainGraph<I>,
+    /// The last active indexes of each keychain
+    pub last_active_indexes: BTreeMap<K, u32>,
+}
+
+impl<K, I> Default for KeychainScan<K, I> {
+    fn default() -> Self {
+        Self {
+            update: Default::default(),
+            last_active_indexes: Default::default(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize, serde::Serialize),
+    serde(
+        crate = "serde_crate",
+        bound(
+            deserialize = "K: Ord + serde::Deserialize<'de>, I: serde::Deserialize<'de>",
+            serialize = "K: Ord + serde::Serialize, I: serde::Serialize"
+        )
+    )
+)]
+pub struct KeychainChangeSet<I, K> {
+    /// The changes that have occured in the blockchain
+    pub chain_graph: chain_graph::ChangeSet<I>,
+    /// The changes in local keychain derivation indicies
+    pub keychain: BTreeMap<K, u32>,
```

Comment:
Should this be `BTreeMap<K, Option<u32>>`,, Like other change set.. Because we could have a change set that doesn't update the last active index?

Reply:
Then there are two ways of representing an not updated keychain: `None` and it not being present. What for?

Reply:
> Then there are two ways of representing an not updated keychain: None and it not being present. What for?

Ya I think you are right.. If its not updated it should not be in the set.. Then my confusion is, why the `ChangeSet` struct has `Option` in it?? Shouldn't the same logic apply there too??

And "if its not updated, its not in the set", makes more intuitive sense to me than having update value as `None`.

Reply:
Which changeset are you talking about here? There is `None` in the `SparseChain`'s changeset because it's possible to delete transactions. Setting it to `None` indicates removal.